File: packages/python/gui/wxogre.py
# Library Includes
import wx
import ogre.renderer.OGRE as Ogre

# Projects Include
from ram.sim.simulation import Simulation
from core import cls_property
import event

class wxOgre(wx.PyControl):
    """
    This control creates and Ogre Rendering window relying on the simulation
    to handle the ogre specific part.
    
    @type camera: Ogre.Camera
    @ivar camera: The camera whos view is shown through the viewport
    """    
    def __init__(self, camera, parent, id = -1, pos = wx.DefaultPosition, 
                 size = wx.DefaultSize, style = 0, validator = wx.DefaultValidator, 
                 name = wx.ControlNameStr):
        wx.PyControl.__init__(self, parent, id, pos, size, style, validator, name)
        
        self._viewport = None
        self._camera = camera
        if camera is not None:
            self._camera.setAutoAspectRatio(True)    
            
        self._init_ogre()
            
    class camera(cls_property):
        """
        The camera attached to the main viewport of the window.
        """
        def fget(self):
            return self._camera
        def fset(self, camera):
            # This ensures the window resizes properly
            camera.setAutoAspectRatio(True)
            
            if self._camera is None:
                self._viewport = self._render_window.addViewport(camera)
            else:
                self._viewport.setCamera(camera)
                
            self._camera = camera
            # Refresh the window
            self._update()
    
    def _init_ogre(self):
        """
        Hook ogre up to the control.  On Linus you must have called "Show()" on 
        your main frame before creating any wxOgre widgets!.
        """
        self._create_ogre_window()
        self._render_window.update()
            
        # Setup our event handlers
        self.Bind(wx.EVT_CLOSE, self._on_close)
        #self.Bind(wx.EVT_ERASE_BACKGROUND, self._update)
        self.Bind(wx.EVT_SIZE, self._update)
            
        self._update()
        
    def _update(self, event = None):
        """
        Handles all events that require redrawing
        """
        # Resize the window on resize
        if type(event) is wx.SizeEvent:
            self._render_window.windowMovedOrResized()         

            if '__WXGTK__' == wx.Platform:
#                size = self.GetClientSize()
                 if self._viewport is not None:
                    self._viewport._updateDimensions()
    
        # Redraw the window for every event
        self._render_window.update()
        
        if event is not None:
            event.Skip()
            
    def _on_close(self, event):
        self._render_window.removeAllViewports()
        self._render_window.destroy()
    
    def _create_ogre_window(self):
        size = self.GetClientSize()
        params = self._get_window_params()
        name = str(self.GetName())
        
        root = Ogre.Root.getSingleton()
        self._render_window = root.createRenderWindow(name, size.width, 
                                                      size.height, False, 
                                                      params)
            
            
        self._render_window.active = True
        
        # You can only create a camera after you have made the first render
        # window, so check to see if we are given a camera
        if self._camera is not None:
            self._viewport = self._render_window.addViewport(self._camera)
            
    def _get_window_params(self):
        """
        Encapsulates the platform specific part of the window creation, grabing
        the windows handle and transforming them for Ogre.
        """
        params = Ogre.NameValuePairList()
        
        if '__WXGTK__' == wx.Platform:
            from gui.util import get_window_handle_str
            params['externalWindowHandle'] = get_window_handle_str(self)
        elif '__WXMSW__' == wx.Platform:
            params['externalWindowHandle'] = str(self.GetHandle())
        else:
            raise Exception('%s no yet supported' % wx.Platform)
        
        return params
    

class wxOgreFrame(wx.Frame):
    """
    This frame will call its on_activate method only once, allowing you to
    properly handle wxWidgets + Ogre initialization in a cross platform manner.
    """
    def __init__(self, parent, id=-1, title = wx.EmptyString, 
                 pos = wx.DefaultPosition, size = wx.DefaultSize, 
                 style = wx.DEFAULT_FRAME_STYLE, name = wx.FrameNameStr):
        wx.Frame.__init__(self, parent, id, title, pos, size, style, name)
        self.Bind(wx.EVT_ACTIVATE, self.on_activate)
        
    def on_activate(self, event):
        self.Bind(wx.EVT_ACTIVATE, None)

File: sandbox/wxogre/src/wxOgre.cpp
#include <iostream>
#include <string>

// wxWidgets includes
#include <wx/event.h>
#include <wx/string.h>
#include <wx/validate.h>
#include <wx/stattext.h>

// Ogre includes
#include <OGRE/OgreRoot.h>
#include <OGRE/OgreRenderWindow.h>
#include <OGRE/OgreRenderSystem.h>
#include <OGRE/OgreStringConverter.h>

// Project includes
#include "wxOgre.h"

// Platform specific includes
#ifdef __WXGTK__
#include <gdk/gdkx.h>
#include <gtk/gtk.h>
#endif

// For the wxWidget RTTI system
IMPLEMENT_DYNAMIC_CLASS(wxOgre, wxControl)

// Sets up callback function for wxWidgets events
BEGIN_EVENT_TABLE(wxOgre, wxControl)
	EVT_SIZE(wxOgre::onSize)
	// EVT_PAINT(wxOgre::OnPaint) // Produces flickers and runs too fast!
	EVT_ERASE_BACKGROUND( wxOgre::onEraseBackground )
END_EVENT_TABLE()

wxOgre::wxOgre() :
    wxControl(),
    mRenderWindow(0),
    mViewport(0),
    mCamera(0),
    mInitialized(false)
{
}

wxOgre::wxOgre(Ogre::Camera* camera, wxWindow* parent, wxWindowID id,
               const wxPoint& pos, const wxSize& size, long style,
               const wxString& name) :
    wxControl(),
    mRenderWindow(0),
    mViewport(0),
    mCamera(0),
    mInitialized(false)
{
    Create(camera, parent, id, pos, size, style, name);
}

bool wxOgre::Create(Ogre::Camera* camera, wxWindow* parent, wxWindowID id,
                    const wxPoint& pos, const wxSize& size, long style,
                    const wxString& name)
{
    mCamera = camera;

    // Create wxWidgets superclass
    wxControl::Create(parent, id, pos, size, style, wxDefaultValidator, name);

    return true;
}

wxOgre::~wxOgre()
{
	// Get rid of all view ports and destroy the render window
    mRenderWindow->removeAllViewports();
	Ogre::Root::getSingleton().detachRenderTarget(mRenderWindow);
	mRenderWindow = 0;
}

void wxOgre::initOgre()
{
    if (!mInitialized)
    {
        mInitialized = true;

        // Create all Ogre objects
        createOgreRenderWindow();
        // Update the render window just once
        update();
    }
}

Ogre::Camera* wxOgre::getCamera(void)
{
    return mViewport->getCamera();
}

void wxOgre::setCamera(Ogre::Camera* camera)
{
    mCamera = camera;

    if (mViewport)
    {
        mViewport->setCamera(camera);
    }
    else if(mInitialized)
    {
        mViewport = mRenderWindow->addViewport(mCamera);
        mViewport->setBackgroundColour(Ogre::ColourValue(1.0f, 0.0f, 0.0f,
                                                         1.0f));
    }
}

Ogre::RenderWindow* wxOgre::getRenderWindow()
{
    return mRenderWindow;
}

bool wxOgre::initialized()
{
    return mInitialized;
}

void wxOgre::createOgreRenderWindow()
{
	// Grab the current render system from Ogre
	Ogre::RenderSystem* renderSystem =
        Ogre::Root::getSingleton().getRenderSystem();

	// Create a new parameters list according to compiled OS
	Ogre::NameValuePairList params;
    getWindowParams(&params);

	// Get wx control window size
	int width;
	int height;
	GetClientSize(&width, &height);
	std::string name(GetName().mb_str(wxConvUTF8));

	// Create the render window (give the name of wxWidget window to Ogre)
	mRenderWindow =
        renderSystem->createRenderWindow(name, width, height, false, &params);

	// Set the viewport up with camera
	// NOTE: You can only create a camera after you have made the first camera
	//       we are going to need to be passed a camera for the second one
	if (mCamera)
	{
        mViewport = mRenderWindow->addViewport(mCamera);
        mViewport->setBackgroundColour(Ogre::ColourValue(1.0f, 0.0f, 0.0f,
                                                         1.0f));
	}
}


void wxOgre::getWindowParams(Ogre::NameValuePairList* params)
{
    Ogre::String handle;
#ifdef __WXMSW__
    mContainer = this;
	handle = Ogre::StringConverter::toString((size_t)((HWND)GetHandle()));
	(*params)["externalWindowHandle"] = handle;

#elif defined(__WXGTK__)
    SetBackgroundStyle(wxBG_STYLE_CUSTOM);
    std::stringstream handleStream;
    GtkWidget* widget =  GetHandle();

    // Display
    Display* display = GDK_WINDOW_XDISPLAY( widget->window );
    handleStream << (unsigned long)display << ':';

    // Screen (returns ":display.screen")
    /* The new Ogre GLX renderer patch removes the need for supplying the
       screen

    std::string screenStr = DisplayString(display);
    int dotpos = screenStr.find(".", 0);
    screenStr = screenStr.substr(dotpos + 1, screenStr.length() - dotpos);
    handleStream << screenStr << ':';*/

    // XID (typedef of an unsigned int)
    Window wid = GDK_WINDOW_XWINDOW( widget->window );
    handleStream << wid;

    handle = handleStream.str();
    (*params)["parentWindowHandle"] = handle;
#else
	#error External/Parent Window handle not supported on this platform
#endif
}

void wxOgre::onSize(wxSizeEvent& event)
{
    if (mInitialized)
    {
        // Setting new size;
        int width;
        int height;

        GetClientSize(&width, &height);
        mRenderWindow->resize( width, height );
        // Letting Ogre know the window has been resized;
        mRenderWindow->windowMovedOrResized();

        // Set the aspect ratio for the new size;
        if (mCamera)
            mCamera->setAspectRatio(Ogre::Real(width) / Ogre::Real(height));

        mViewport->_updateDimensions();
        update();
    }
}

void wxOgre::onPaint(wxPaintEvent& event)
{
//	update(); // Produces flickers and runs too fast!
}
void wxOgre::onEraseBackground( wxEraseEvent& )
{
	update();
}

void wxOgre::update()
{
    if (mInitialized)
    {
        // ****************************************************
        // TODO: REMOVE THESE LINES! These are merely for test!
        static float redTone = 0;
        redTone += 0.01;
        if(redTone>1.0)
            redTone=0;
        // ****************************************************

        if(mViewport)
            mViewport->setBackgroundColour(Ogre::ColourValue(redTone, 0.0f, 0.0f, 1.0f));
        mRenderWindow->update();
        std::cout << "Updated" << std::endl;
    }
}
